File: claude/backend/src/main/java/com/ecommerce/service/CarrinhoService.java
package com.ecommerce.service;

import com.ecommerce.domain.Carrinho;
import com.ecommerce.domain.Produto;
import com.ecommerce.domain.Usuario;
import com.ecommerce.domain.exception.RecursoNaoEncontradoException;
import com.ecommerce.repository.CarrinhoRepository;
import com.ecommerce.repository.ProdutoRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Casos de uso do carrinho de compras (RF11-RF15). O service orquestra; as
 * regras (validacao de estoque, calculo de total) ficam no agregado Carrinho.
 */
@Service
@Transactional
public class CarrinhoService {

    private final CarrinhoRepository carrinhoRepository;
    private final ProdutoRepository produtoRepository;

    public CarrinhoService(CarrinhoRepository carrinhoRepository, ProdutoRepository produtoRepository) {
        this.carrinhoRepository = carrinhoRepository;
        this.produtoRepository = produtoRepository;
    }

    /** Obtem o carrinho ativo do usuario, criando-o caso ainda nao exista. */
    public Carrinho obterCarrinho(Usuario usuario) {
        return carrinhoRepository.findByUsuarioId(usuario.getId())
                .orElseGet(() -> carrinhoRepository.save(new Carrinho(usuario.getId())));
    }

    /** Adiciona um item ao carrinho (RF11). */
    public Carrinho adicionarItem(Usuario usuario, Long produtoId, Integer quantidade) {
        Carrinho carrinho = obterCarrinho(usuario);
        Produto produto = buscarProdutoAtivo(produtoId);
        carrinho.adicionarProduto(produto, quantidade);
        return carrinhoRepository.save(carrinho);
    }

    /** Altera a quantidade de um item (RF12). */
    public Carrinho alterarQuantidade(Usuario usuario, Long produtoId, Integer quantidade) {
        Carrinho carrinho = obterCarrinho(usuario);
        Produto produto = buscarProdutoAtivo(produtoId);
        carrinho.alterarQuantidade(produto, quantidade);
        return carrinhoRepository.save(carrinho);
    }

    /** Remove um item do carrinho (RF13). */
    public Carrinho removerItem(Usuario usuario, Long produtoId) {
        Carrinho carrinho = obterCarrinho(usuario);
        Produto produto = buscarProdutoAtivo(produtoId);
        carrinho.removerProduto(produto);
        return carrinhoRepository.save(carrinho);
    }

    private Produto buscarProdutoAtivo(Long produtoId) {
        return produtoRepository.findById(produtoId)
                .filter(Produto::isAtivo)
                .orElseThrow(() -> new RecursoNaoEncontradoException("Produto nao encontrado."));
    }
}

File: codex/src/application/services/carrinho-service.ts
import { Carrinho } from "../../domain/entities/carrinho.js";
import { ApplicationError } from "../errors/application-error.js";
import { Repositories } from "../repositories.js";

export class CarrinhoService {
  constructor(private readonly repositories: Repositories) {}

  async obterCarrinho(usuarioId: number): Promise<Carrinho> {
    const carrinho = await this.repositories.carrinhos.buscarPorUsuarioId(usuarioId);
    if (carrinho) return carrinho;

    const novoCarrinho = new Carrinho({ id: this.repositories.carrinhos.proximoId(), usuarioId });
    await this.repositories.carrinhos.salvar(novoCarrinho);
    return novoCarrinho;
  }

  async adicionarItem(usuarioId: number, produtoId: number, quantidade: number): Promise<Carrinho> {
    const [carrinho, produto] = await Promise.all([
      this.obterCarrinho(usuarioId),
      this.repositories.produtos.buscarPorId(produtoId)
    ]);
    if (!produto || !produto.ativo) throw new ApplicationError("Produto nao encontrado.", 404);

    carrinho.adicionarProduto(produto, quantidade);
    await this.repositories.carrinhos.salvar(carrinho);
    return carrinho;
  }

  async alterarItem(usuarioId: number, produtoId: number, quantidade: number): Promise<Carrinho> {
    const [carrinho, produto] = await Promise.all([
      this.obterCarrinho(usuarioId),
      this.repositories.produtos.buscarPorId(produtoId)
    ]);
    if (!produto || !produto.ativo) throw new ApplicationError("Produto nao encontrado.", 404);

    carrinho.alterarQuantidade(produto, quantidade);
    await this.repositories.carrinhos.salvar(carrinho);
    return carrinho;
  }

  async removerItem(usuarioId: number, produtoId: number): Promise<Carrinho> {
    const [carrinho, produto] = await Promise.all([
      this.obterCarrinho(usuarioId),
      this.repositories.produtos.buscarPorId(produtoId)
    ]);
    if (!produto) throw new ApplicationError("Produto nao encontrado.", 404);

    carrinho.removerProduto(produto);
    await this.repositories.carrinhos.salvar(carrinho);
    return carrinho;
  }

  async limpar(usuarioId: number): Promise<Carrinho> {
    const carrinho = await this.obterCarrinho(usuarioId);
    carrinho.limpar();
    await this.repositories.carrinhos.salvar(carrinho);
    return carrinho;
  }
}
